vsphere/datadog_checks/vsphere/metrics.py

```diff
@@ -78,36 +78,36 @@
 
 # All metrics that can be collected from VirtualMachines.
 # The table maps a dd-formatted metric_name to a tuple containing:
-# (collection_level, per_instance_collection_level, (optional)is_available_per_instance)
+# (collection_level, per_instance_collection_level, is_available_per_instance)
 # Note: Only the keys of the dict are used, but the values can be useful later
 # to decide how and when to collect per-instance level metrics.
 VM_METRICS = {
-    'cpu.costop.sum': (2, 3),
-    'cpu.demand.avg': (2, 3),
-    'cpu.demandEntitlementRatio.latest': (4, 4),
-    'cpu.entitlement.latest': (2, 3),
+    'cpu.costop.sum': (2, 3, False),
+    'cpu.demand.avg': (2, 3, False),
+    'cpu.demandEntitlementRatio.latest': (4, 4, False),
+    'cpu.entitlement.latest': (2, 3, False),
     'cpu.idle.sum': (2, 3, True),
-    'cpu.latency.avg': (2, 3),
+    'cpu.latency.avg': (2, 3, False),
     'cpu.maxlimited.sum': (2, 3, True),
     'cpu.overlap.sum': (3, 3, True),
-    'cpu.readiness.avg': (4, 4),
-    'cpu.ready.sum': (1, 3),
+    'cpu.readiness.avg': (4, 4, False),
+    'cpu.ready.sum': (1, 3, False),
     'cpu.run.sum': (2, 3, True),
-    'cpu.swapwait.sum': (3, 3),
+    'cpu.swapwait.sum': (3, 3, False),
     'cpu.system.sum': (3, 3, True),
     'cpu.usage.avg': (1, 3, True),
```

Comment:
Unfortunately this should be false, `cpu.usage.avg` has no per-instance value when looking at VMs.

This leads to the caveat that `vsphere.cpu.usage.avg` is not emitted anymore for VMs.

I think we'll need to make a new pass on all those metrics to check the validity of this boolean.

Reply:
docs are not really helpful though: https://code.vmware.com/apis/704/vsphere/cpu_counters.html

Reply:
Sure, will double check all the metrics then ;)

Reply:
Instead of reverse engineering the per-instance bool/state, I think it might worth using the logic I proposed previously. What do you think ? That will avoid having to maintain a hard coded status.

Basically, for a given metric+resource_type, if one of the metric have an instance tag, we consider that this metric+resource_type is a per-instance metric.

```
        have_instance_value = defaultdict(set)

        for results_per_mor in query_results:
            resource_type = type(results_per_mor.entity)
            metadata = self.metrics_metadata_cache.get_metadata(resource_type)
            for result in results_per_mor.value:
                metric_name = metadata.get(result.id.counterId)
                if result.id.instance:
                    have_instance_value[resource_type].add(metric_name)
        # usage
       if have_instance_value[resource_type][metric_name]:
          ....
```

Reply:
I think I missed that idea before, but I agree it's definitely the best way to go. Let's do this 👍

Reply:
We can remove the `per-instance` field in `metrics.py` then.
Also, we may be able to change everything in `metrics.py` to a set (just discard the values of the dictionaries).
The `collection_level` is not used and won't be, when listing metrics from the vSphere API, we already specify the collection level.
The `per_instance_collection_level` is not used and won't be, as we rely on a explicit configuration from the users.
The `is_available_per_instance` is not used and won't be with this logic.